stdlib/public/core/StringObject.swift

```diff
@@ -23,6 +19,139 @@
 
 @_fixed_layout @usableFromInline
 internal struct _StringObject {
+  /*
+
+  On 64-bit platforms, the discriminator is the most significant 8 bits of the
+  bridge object.
+
+  ┌───────────────────╥─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┐
+  │ Form              ║  7  │  6  │  5  │  4  │  3  │  2  │  1  │  0  │
+  ╞═══════════════════╬═════╪═════╪═════╪═════╪═════╧═════╧═════╧═════╡
+  │ Immortal, Small   ║  1  │ASCII│  1  │  0  │      small count      │
+  ├───────────────────╫─────┼─────┼─────┼─────┼─────┬─────┬─────┬─────┤
+  │ Immortal, Large   ║  1  │  0  │  0  │  0  │  0  │ TBD │ TBD │ TBD │
+  ╞═══════════════════╬═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╡
+  │ Native            ║  0  │  0  │  0  │  0  │  0  │ TBD │ TBD │ TBD │
+  ├───────────────────╫─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┤
+  │ Shared            ║  0  │  0  │  0  │  0  │  1  │ TBD │ TBD │ TBD │
+  ├───────────────────╫─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┤
+  │ Exotic            ║  0  │  0  │  0  │  1  │  1  │ TBD │ TBD │ TBD │
+  ╞═══════════════════╬═════╪═════╪═════╪═════╪═════╪═════╪═════╪═════╡
+  │ ObjC, Contiguous  ║  0  │  1  │  0  │  0  │  1  │ TBD │ TBD │ TBD │
+  ├───────────────────╫─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┤
+  │ ObjC, Slow        ║  0  │  1  │  0  │  1  │  1  │ TBD │ TBD │ TBD │
+  └───────────────────╨─────┴─────┴─────┴─────┴─────┴─────┴─────┴─────┘
+
+  b3: isForeign (i.e., object payload isn't a biased pointer)
+  b4: isSlow
+  b5: isSmall
+  b6: isObjC/isASCII
+  b7: isImmortal
+
+  On 32-bit platforms, we use a 7-bit discriminator with the same encoding as
+  above, except bit 7 is omitted from storage, so that we supply extra
+  inhabitants in the String structure. Bit 7 can be recovered by looking at
+  `_variant`.
+
+  */
+  @_fixed_layout @usableFromInline
+  struct Discriminator {
+    @usableFromInline
+    internal var _value: UInt8
+
+    @inlinable @inline(__always)
+    internal init(_ value: UInt8) {
+      self._value = value
+    }
+  }
+
+#if arch(i386) || arch(arm)
+  @usableFromInline @_frozen
+  internal enum Variant {
+    case immortal(bitPattern: UInt)
+    case native(storage: _AbstractStringStorage)
+    case objC(object: _CocoaString)
+
+    @inlinable @inline(__always)
+    internal static func immortal(start: UnsafePointer<UInt8>) -> Variant {
+      let biased = UInt(bitPattern: start) &- _StringObject.nativeBias
+      return .immortal(bitPattern: biased)
+    }
+
+    @inlinable
+    internal var isImmortal: Bool {
+      @inline(__always) get {
+        if case .immortal = self { return true }
+        return false
+      }
+    }
+  }
+
+  @_fixed_layout @usableFromInline
+  struct Flags {
+    @usableFromInline
+    internal var _value: UInt16
+
+    @inlinable @inline(__always)
+    init(_ value: UInt16) {
+      self._value = value
+    }
+  }
+
+  @_fixed_layout @usableFromInline
+  struct CountAndFlags {
```

Comment:
This struct is defined purely for consistency with 64-bit archs; it is only used in storage classes.